mozregression/main.py

```diff
@@ -308,6 +308,89 @@ def check_mozregression_version():
         )
 
 
+def check_unprivileged_userns():
+    """
+    Some distribution started to block unprivileged user namespaces via
+    AppArmor.  This might result in crashes on older builds, and in degraded
+    sandbox behavior.  It is fixed with an AppArmor profile that allows the
+    syscall to proceed, but this is path dependant on the binary we download
+    and needs to be installed at a system level, so we can only advise people
+    of the situation.
+
+    The following sys entry should be enough to verify whether it is blocked or
+    not, but the Ubuntu security team recommend cross-checking with actual
+    syscall.  This code is a simplification of how Firerox does it, cf
+    https://searchfox.org/mozilla-central/rev/23efe2c8c5b3a3182d449211ff9036fb34fe0219/security/sandbox/linux/SandboxInfo.cpp#114-175
+    and has been the most reliable way so far (shell with unshare would not
+    reproduce EPERM like we want).
+
+    We need to build it (via g++) and then run it. It will just return 0 (exit
+    code) but it will print via the child process either "0" if it was OK, orxi
+    "-1:X" with X being the errno. We are interested in errno being EPERM which
+    shows the syscall is blocked.
+    """
+
+    apparmor_file = "/proc/sys/kernel/apparmor_restrict_unprivileged_userns"
+    if not os.path.isfile(apparmor_file):
+        return
+
+    with open(apparmor_file, "r") as f:
+        if f.read().strip() != "1":
+            return
+
+    TEST_PROGRAM = """
+#include <cstdlib>
+#include <cstdio>
+#include <cerrno>
+
+#include <sys/syscall.h>
+#include <unistd.h>
+
+#include <signal.h>
+
+#if !defined(_GNU_SOURCE)
+#define _GNU_SOURCE
+#endif
+#include <sched.h>
+
+int main() {
+  pid_t pid = syscall(SYS_clone, SIGCHLD | CLONE_NEWUSER, nullptr, nullptr, nullptr, nullptr);
+  if (pid == 0) {
+    // child side
+    int rv = unshare(CLONE_NEWPID);
+    printf("%d", rv);
+    if (rv < 0) {
+      printf(":%d", errno);
+    }
+  }
+  exit(0);
+}
+"""
+
+    import subprocess
+    import tempfile
+
+    LOG.warning("Unprivileged user namespaces might be disabled. Checking (requires g++) ...")
+
+    with tempfile.NamedTemporaryFile(mode="w", suffix=".cc") as source:
```

Comment:
So as mentionned in the comment, this will run only on system where it's mostly certain we are blocked, but it was a suggestion from ubuntu security team that this was the most reliable way. I'm still looking for some `unshare` call that would replicate. If i cannot find one and we decide we really dont want such a flow of building code there, then we can drop it at the expense of potentially a useless warning to some users (but most of the time the warning should be correct).